File: problems/leetcode/0600/658/658.py
# 658. Find K Closest Elements
# https://leetcode.com/problems/find-k-closest-elements/description/

# 1. Naive Solution
# 주어진 정렬된 배열에서 x의 lower bound와 그 바로 전 원소 중 하나가 가장 가까운 원소
# 그 원소를 기준으로 Two Pointer를 이용해서 k개를 선정
# T(logn + k)

# 2. Solution
# Lower Bound를 직접 Binary Search로 찾기

from typing import *

class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        n = len(arr)
        left, right = 0, n - k

        while left < right:
            mid = (left + right) // 2

            if x - arr[mid] > arr[mid + k] - x:
                left = mid + 1
            else:
                right = mid

        return arr[left:left + k]

File: problems/leetcode/0600/658/658.cpp
// 658. Find K Closest Elements
// https://leetcode.com/problems/find-k-closest-elements/

using namespace std;

#include <vector>

class Solution {
public:
  vector<int> findClosestElements(vector<int> &arr, int k, int x) {
    int mid = lower_bound(arr.begin(), arr.end(), x) - arr.begin();
    if (mid >= arr.size() || mid > 0 && arr[mid] - x >= x - arr[mid - 1]) {
      mid -= 1;
    }

    int left = mid - 1, right = mid + 1;
    for (int i = 0; i < k - 1; i++) {
      if (left < 0) {
        right++;
        continue;
      }

      if (right >= arr.size()) {
        left--;
        continue;
      }

      if (x - arr[left] <= arr[right] - x) {
        left--;
      } else {
        right++;
      }
    }

    return vector<int>(arr.begin() + left + 1, arr.begin() + right);
  }
};
